pkg/apply/mutator/apply-time-mutator.go

```diff
@@ -0,0 +1,216 @@
+
+package mutator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/cli-utils/pkg/apply/cache"
+	"sigs.k8s.io/cli-utils/pkg/kyq"
+	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/cli-utils/pkg/object/mutation"
+)
+
+// ApplyTimeMutator mutates a resource by injecting values specified by the
+// apply-time-mutation annotation.
+// The optional ResourceCache will be used to speed up source reosurce lookups,
+// if specified.
+// Implements the Mutator interface
+type ApplyTimeMutator struct {
+	Client        dynamic.Interface
+	Mapper        meta.RESTMapper
+	ResourceCache cache.ResourceCache
+}
+
+// Name returns a mutator identifier for logging.
+func (atm *ApplyTimeMutator) Name() string {
+	return "ApplyTimeMutator"
+}
+
+// Mutate parses the apply-time-mutation annotation and loops through the
+// substitutions, applying each of them to the supplied target object.
+// Returns true with a reason, if mutation was performed.
+func (atm *ApplyTimeMutator) Mutate(obj *unstructured.Unstructured) (bool, string, error) {
+	mutated := false
+	reason := ""
+
+	if !mutation.HasAnnotation(obj) {
+		return mutated, reason, nil
+	}
+
+	subs, err := mutation.ReadAnnotation(obj)
+	if err != nil {
+		return mutated, reason, errors.Wrapf(err, "failed to read jsonpath field in target resource: %v", mutation.NewResourceReference(obj))
+	}
+
+	targetID := object.UnstructuredToObjMetaOrDie(obj)
+	klog.V(5).Infof("target resource %s/%s: %#v", targetID.Namespace, targetID.Name, obj)
+
+	for _, sub := range subs {
+		sourceRef := sub.SourceRef
+
+		// lookup source resource from cache or cluster
+		sourceObj, err := atm.getObject(sourceRef)
+		if err != nil {
+			return mutated, reason, errors.Wrapf(err, "failed to retrieve resource from sourceRef: %v", sourceRef)
+		}
+
+		sourceID := object.UnstructuredToObjMetaOrDie(sourceObj)
+		klog.V(5).Infof("source resource %s/%s: %#v", sourceID.Namespace, sourceID.Name, sourceObj)
+
+		// lookup target field in target resource
+		targetValue, err := readFieldValue(obj, sub.TargetPath)
+		if err != nil {
+			return mutated, reason, errors.Wrapf(err, "failed to read field from target resource: %v", mutation.NewResourceReference(obj))
+		}
+
+		// lookup source field in source resource
+		sourceValue, err := readFieldValue(sourceObj, sub.SourcePath)
+		if err != nil {
+			return mutated, reason, errors.Wrapf(err, "failed to read field from source resource: %v", sourceRef)
+		}
+
+		var newValue interface{}
+		if sub.Token == "" {
+			// token not specified, replace the entire target value with the source value
+			newValue = sourceValue
+		} else {
+			// token specified, substitute token for source field value in target field value
+			targetValueString, ok := targetValue.(string)
```

Comment:
wdyt about doing some type assertions for templated destinations,?
they should probably all be strings... or just let it fall through

Reply:
The impl here errors if the token is specified and the target field isn't a string.

Having token be optional allows for maps, lists, ints, floats, and bools to be set wholesale, without string substitution.